pcsx2-qt/Debugger/DataInspectorWindow.cpp

```diff
@@ -0,0 +1,202 @@
+#include "DataInspectorWindow.h"
+
+#include "../QtHost.h"
+#include "common/Error.h"
+#include "CDVD/CDVD.h"
+#include "DebugTools/ccc/analysis.h"
+#include "Debugger/Delegates/DataInspectorValueColumnDelegate.h"
+
+DataInspectorWindow::DataInspectorWindow(QWidget* parent)
+	: QMainWindow(parent)
+{
+	m_ui.setupUi(this);
+
+	// probably a race condition
+	std::string elfPath = VMManager::GetELFPath();
```

Comment:
Comment's correct, this is a race. This whole bit is - CDVD functions cannot be called anywhere other than the CPU thread.

Probably should do it on the CPU thread at ELF load time, like the current symbol stuff.

Reply:
I'm not sure what the best thing to for that is currently. The basic problem is that the AST in memory can be quite large (e.g. about 250MB in memory for The Sims, there are probably worse examples), so I didn't want it to be loaded if the data inspector wasn't opened. I could either load the symbols all the time, keep the unparsed symbol table in memory all the time (which is much smaller), or rework the ELF code so I can do what I'm doing now, but actually thread safe. In the former case I could see what I could do to make the AST smaller in memory. There are some trivial things I can think of, and some less trivial things.

Reply:
Then it should kick a request off to the CPU thread asynchronously, and pass it back to the UI thread. You can use `Host::RunOnCPUThread()` and `QtHost::RunOnUIThread()` to do so without creating additional entry points, just have to be careful if the window gets closed before the UI callback fires.

Reply:
Check the new version of the code. I've put the CDVD code on the CPU thread and spun off the symbol table parsing onto its own thread using QtConcurrent::run. Also, I went in with massif and got CCC's memory usage down by over 50%.